### ISSUE TITLE
Less search results with active listenport than in SoulseekQt? #question

### ISSUE THREAD

**Issue by nzsb** [jpdillingham/Soulseek.NET#353]
Hi,

Thanks for the great work, the library is overall good.
I am just wondering why it could be that the library returns less search results as the SoulseekQt client? I am using both with active NAT-forwarded listening ports. With the library I checked the logs, the client is accepting incoming requests. I am using default timeouts, in the SoulseekQt client the more results are coming back <1 second.
Sometimes it leads to no results which is why I noticed this and I am wondering if I am doing something wrong.

Thanks!

**Comment by jpdillingham** [comment]
Thanks for raising this.  I did some investigation last night and discovered that locked files are being excluded from both browse and search results.  I've created issue #354 to track this and have begun working on an implementation.

Can you take a look and confirm that the missing files are locked? If not, please provide some example search terms so I can investigate further.

In general you can expect the same results between the two for the same search term, less any users that may have gone offline between the searches.

Also, note that by default results which don't have an available download slot are hidden.

**Comment by nzsb** [comment]
Thanks for the quick response!
Locked files I did not notice, but it is very marginal for most of my searches and I think it is not an issues at all (for me at least :-))
I did some research on this, it seems it has something to do with non-alphanumeric characters, but I am not sure.
My test queries (in QT using without "") are below. QT results are all non-locked files and users with free slots. In QT I waited around 3 seconds for the results.
In C# client I use:  _client.SearchAsync(SearchQuery.FromText(title), options:new SearchOptions());


"Huey Lewis And The News - The Power Of Love":
C#: 0
QT: >100

if I remove the dash it is much better:

"Huey Lewis And The News The Power Of Love":
C#: 100
QT: >100

"wham! - wake me up before you go-go":
C#: 0
QT: > 10

if I remove the exclamation mark, it is much better:

"wham wake me up before you go-go":
C#: 20
QT: 26

"Usher - Dj Got Us Fallin' In Love (Feat. Pitbull)":
C#: 0
QT: 2

if I remove the parenthenses, it is much better:

"Usher - Dj Got Us Fallin' In Love Feat. Pitbull":
C#: 7
QT: >10

I hope this helps.

**Comment by jpdillingham** [comment]
Thanks, your feedback is super helpful.

It looks like Qt is transforming input before sending it to the server, whereas I send the text as entered (at least, that was my intent).  I'll need to spend some time with Wireshark to figure out where the discrepancy is.

**Comment by nzsb** [comment]
Sounds reasonable. You are really hardcore with this reverse-engineering. Thanks for looking into this!

**Comment by jpdillingham** [comment]
I believe I've discovered an issue (well, two: #357)

It doesn't look like QT modifies search terms in any way when sending them to the server, but for some reason it repeats them.  Both this library and QT send the following message given the input "george michael wham wake me up before you go-go":

```
      <length:59> <code:26>   <token>     <strlen>    <search text>
[lib] 3b 00 00 00 1a 00 00 00 XX XX XX XX 2f 00 00 00 67 65 6f 72 67 65 20 6d 69 63 68 61 65 6c 20 77 68 61 6d 20 77 61 6b 65 20 6d 65 20 75 70 20 62 65 66 6f 72 65 20 79 6f 75 20 67 6f 2d 67 6f
[qt]  3b 00 00 00 1a 00 00 00 XX XX XX XX 2f 00 00 00 67 65 6f 72 67 65 20 6d 69 63 68 61 65 6c 20 77 68 61 6d 20 77 61 6b 65 20 6d 65 20 75 70 20 62 65 66 6f 72 65 20 79 6f 75 20 67 6f 2d 67 6f
```

QT appends an additional message:

```
<length:55> <code:153>              <strlen>    <search text>
37 00 00 00 99 00 00 00             2f 00 00 00 67 65 6f 72 67 65 20 6d 69 63 68 61 65 6c 20 77 68 61 6d 20 77 61 6b 65 20 6d 65 20 75 70 20 62 65 66 6f 72 65 20 79 6f 75 20 67 6f 2d 67 6f
```

Code `153` is not described in any available documentation, so I'm really not sure what this is.  The search text is identical, but the message doesn't include the `token`, which is what other clients need to use when responding to the search so the results can be correlated properly.

I'll keep playing around with this to see if I can infer the significance of this message, I just wanted to update in case anyone stumbles upon this later.

**Comment by jpdillingham** [comment]
Upon further testing, QT _does_ manipulate the string; it removes `-`, but only if it isn't adjacent to another word, so `go-go` is unaffected but `News - The` becomes `News The`.

`Usher - Dj Got Us Fallin' In Love (Feat. Pitbull)` becomes `Usher Dj Got Us Fallin' In Love (Feat. Pitbull)`, but the second (still unknown) message contains the unmodified text.

All the combinations I'm trying so far seem to point to `-` with spaces on either side being dropped, and that seems to be the only modification.

It doesn't seem like the second search message impacts the results.

**Comment by nzsb** [comment]
So QT does not replace so much after all.
If in my client I remove all punctuation characters before searching, clients can do a lot better job at searching, and I get a lot better results than normally in QT, which is great.

"Usher - Dj Got Us Fallin' In Love (Feat. Pitbull)"
C#: 0 response
QT: 3 responses
(the case you discovered)

"Usher Dj Got Us Fallin' In Love (Feat. Pitbull)"
C#: 1 response
QT: 3 responses
(although here I am still wondering about the difference)

"Usher Dj Got Us Fallin In Love (Feat. Pitbull)"
C#: 1 response
QT: 2 responses

"Usher  Dj Got Us Fallin In Love Feat Pitbull"
C#: 100 responses
QT: >100

**Comment by mathiascode** [comment]
Nicotine+ developer here, chiming in on this issue with my discoveries.

Code 153 seems to be for receiving related searches, which shows up as a separate pane next to the search results in SoulseekQt. I'm not exactly sure how this works on the server side though, since I've barely seen any related searches. https://github.com/Nicotine-Plus/nicotine-plus/blob/master/doc/SLSKPROTOCOL.md#server-code-153

Something I noticed while improving searches in Nic+, is that SoulseekQt clients don't seem to return any results if a search term includes special characters. I don't know why this is the case, but I added an option to automatically strip out all special characters from search terms.

**Comment by jpdillingham** [comment]
@mathiascode thanks for chiming in! I doubt I would have ever figured this out, as I don't recall ever seeing a related search tab.

Good to know about stripping characters; I'll do a little more investigating to see if I can figure out a definitive list.

**Comment by jpdillingham** [comment]
@nzsb I've just released an update ([1.0.0-rc3](https://github.com/jpdillingham/Soulseek.NET/releases/tag/1.0.0-rc3)) that has greatly improved the parity in results with QT.

I've done a bit of testing this morning and the results are within a few responses of one another, which I think is reasonable given the nature of the distributed network.

Please have a look and let me know!

**Comment by nzsb** [comment]
Thanks, I did some testing, results are really improved, but for "Huey Lewis And The News - The Power Of Love" the library returns only 3 results, while QT around 100.

To be honest what I did in my code is that I just remove all special characters from the search string and I get lot better results even than QT, so I am happy with that.
